File: src/content/docs/topics/cs/coding-problems/2d-dynamic-programming/494-target-sum-approach3.go
package main

import "fmt"

func findTargetSumWays(nums []int, target int) int {
	total := 0
	for _, x := range nums {
		total += x
	}
	sum := total + target
	if sum < 0 || sum%2 != 0 {
		return 0
	}
	if total < abs(target) {
		return 0
	}
	P := sum / 2
	dp := make([]int, P+1)
	dp[0] = 1
	for _, x := range nums {
		for s := P; s >= x; s-- {
			dp[s] += dp[s-x]
		}
	}
	return dp[P]
}

func abs(x int) int {
	if x < 0 {
		return -x
	}
	return x
}

func main() {
	assert := func(cond bool) {
		if !cond {
			panic("assertion failed")
		}
	}
	assert(findTargetSumWays([]int{1, 1, 1, 1, 1}, 3) == 5)
	assert(findTargetSumWays([]int{1}, 1) == 1)
	assert(findTargetSumWays([]int{1, 1}, 0) == 2)
	assert(findTargetSumWays([]int{1, 2}, 4) == 0)
	assert(findTargetSumWays([]int{1}, -1) == 1)
	assert(findTargetSumWays([]int{0, 0, 0}, 0) == 8)
	fmt.Println("all tests pass")
}

File: src/content/docs/topics/cs/coding-problems/2d-dynamic-programming/494-target-sum-approach3.py
def find_target_sum_ways(nums: list[int], target: int) -> int:
    total = sum(nums)
    if (total + target) % 2 or total < abs(target):
        return 0
    P = (total + target) // 2
    dp = [0] * (P + 1)
    dp[0] = 1
    for x in nums:
        for s in range(P, x - 1, -1):
            dp[s] += dp[s - x]
    return dp[P]

assert find_target_sum_ways([1, 1, 1, 1, 1], 3) == 5
assert find_target_sum_ways([1], 1) == 1
assert find_target_sum_ways([1, 1], 0) == 2
assert find_target_sum_ways([1, 2], 4) == 0
assert find_target_sum_ways([1], -1) == 1
assert find_target_sum_ways([0, 0, 0], 0) == 8
print("all tests pass")
